File: Sources/CountriesFlagsObjC/source/Extensions/UIBezierPath+Star.m
//
//  UIBezierPath+Star.m
//  CountriesFlags
//

#import "Extensions/UIBezierPath+Star.h"

@implementation UIBezierPath (Star)

+ (UIBezierPath *)starAtPoint:(CGPoint)point
                       radius:(CGFloat)radius
                         type:(FlagStarTypes)type {
    UIBezierPath* path = [UIBezierPath bezierPath];
    
    [path moveToPoint:CGPointMake(point.x, -radius + point.y)];
    
    switch (type) {
        case FlagStarTypeFourPointed: {
            [self fourPointerStarAtPoint:point
                                  radius:radius
                             innerRadius:radius * 2.0f / 6.0f
                                    path:path];
        }
            break;
            
        case FlagStarTypeFivePointed: {
            [self fivePointerStarAtPoint:point radius:radius path:path];
        }
            break;
            
        case FlagStarTypeSixPointed: {
            [self sixPointerStarAtPoint:point radius:radius path:path];
        }
            break;
            
        case FlagStarTypeSevenPointed: {
            [self sevenPointerStarAtPoint:point radius:radius path:path];
        }
            break;
            
        case FlagStarTypeEightPointed: {
            [self eightPointerStarAtPoint:point radius:radius path:path];
        }
            break;
            
        case FlagStarTypeTwelvePointed: {
            [self twelvePointerStarAtPoint:point
                                    radius:radius
                               innerRadius:radius / 2.0f
                                      path:path];
        }
            break;
            
        case FlagStarTypeFourteenPointed: {
            [self fourteenPointerStarAtPoint:point radius:radius path:path];
        }
            break;
            
        case FlagStarTypeTwentyFourPointed: {
            [self twentyFourPointerStarAtPoint:point
                                        radius:radius
                                   innerRadius:radius / 2.0f
                                          path:path];
        }
            break;
            
        default:
            // TODO: throw exception?
            break;
    }
    
    [path closePath];
    return path;
}

+ (UIBezierPath *)starAtPoint:(CGPoint)point
                       radius:(CGFloat)radius
                  innerRadius:(CGFloat)innerRadius
                         type:(FlagStarTypes)type {
    UIBezierPath* path = [UIBezierPath bezierPath];
    
    [path moveToPoint:CGPointMake(point.x, -radius + point.y)];
    
    switch (type) {
        case FlagStarTypeFourPointed: {
            [self fourPointerStarAtPoint:point
                                  radius:radius
                             innerRadius:innerRadius
                                    path:path];
        }
            break;

        case FlagStarTypeTwelvePointed: {
            [self twelvePointerStarAtPoint:point
                                    radius:radius
                               innerRadius:innerRadius
                                      path:path];
        }
            break;
            
        case FlagStarTypeTwentyFourPointed: {
            [self twentyFourPointerStarAtPoint:point
                                        radius:radius
                                   innerRadius:innerRadius
                                          path:path];
        }
            break;
            
        default:
            // TODO: throw exception?
            break;
    }
    
    [path closePath];
    return path;
}

+ (void)fourPointerStarAtPoint:(CGPoint)point
                        radius:(CGFloat)radius
                   innerRadius:(CGFloat)innerRadius
                          path:(UIBezierPath *)path {
    for (int index = 1; index < 8; index++) {
        CGFloat r = index % 2 == 0 ? radius : innerRadius;
        double angle = -M_PI / 2.0 + index * M_PI / 4.0;

        double x = r * cos(angle);
        double y = r * sin(angle);
        [path addLineToPoint:CGPointMake(point.x + x, point.y + y)];
    }
}

+ (void)fivePointerStarAtPoint:(CGPoint)point
                        radius:(CGFloat)radius
                          path:(UIBezierPath *)path {
    double theta = 2.0 * M_PI * 2.0f / 5.0f;
    for (int i = 1; i < 5; i++) {
        double x = radius * sin(i * theta);
        double y = radius * cos(i * theta);
        [path addLineToPoint:CGPointMake(point.x + x, point.y - y)];
//         NSLog(@"Five pointed star: x: %f, y: %f", x, y);
    }
}

+ (void)sixPointerStarAtPoint:(CGPoint)point
                       radius:(CGFloat)radius
                         path:(UIBezierPath *)path {
    CGFloat angle = -M_PI / 2.0f;
    for (int index = 0; index < 6; index++) {
        CGFloat xPoint = point.x + cos(angle) * radius;
        CGFloat yPoint = point.y + sin(angle) * radius;
        if (index == 0) {
            [path moveToPoint:CGPointMake(xPoint, yPoint)];
        } else {
            [path addLineToPoint:CGPointMake(xPoint, yPoint)];
        }
        angle += M_PI / 6.0f;

        [path addLineToPoint:CGPointMake(point.x + cos(angle) * (radius * 0.58),
                                         point.y + sin(angle) * (radius * 0.58))];
        angle += M_PI / 6.0f;
    }
}

+ (void)sevenPointerStarAtPoint:(CGPoint)point
                         radius:(CGFloat)radius
                           path:(UIBezierPath *)path {
    CGFloat innerRadius = radius / 2.0f * 0.7f;
    CGFloat angleIncrement = M_PI * 2.0f / 7.0f;
    
    CGFloat startAngle = -M_PI / 2.0;
    [path moveToPoint:CGPointMake(point.x + radius * cos(startAngle),
                                  point.y + radius * sin(startAngle))];
    
    for (int index = 1; index < 14; index++) {
        CGFloat angle = startAngle + angleIncrement * index;
        CGFloat currentRadius = index % 2 == 0 ? radius : innerRadius;
        
        [path addLineToPoint:CGPointMake(point.x + currentRadius * cos(angle),
                                         point.y + currentRadius * sin(angle))];
    }
}

+ (void)eightPointerStarAtPoint:(CGPoint)point
                         radius:(CGFloat)radius
                           path:(UIBezierPath *)path {
    for (int index = 0; index < 16; index++) {
        CGFloat angle = index * M_PI / 8.0;
        CGFloat r = index % 2 == 0 ? radius : radius / 2.0f;
        if (index == 0) {
            [path moveToPoint:CGPointMake(point.x + r * cos(angle),
                                          point.y + r * sin(angle))];
        } else {
            [path addLineToPoint:CGPointMake(point.x + r * cos(angle),
                                             point.y + r * sin(angle))];
        }
    }
}

+ (void)twelvePointerStarAtPoint:(CGPoint)point
                          radius:(CGFloat)radius
                     innerRadius:(CGFloat)innerRadius
                            path:(UIBezierPath *)path {
    [path moveToPoint:CGPointMake(point.x + radius * cos(0.0f),
                                  point.y + radius * sin(0.0f))];

    for (int index = 1; index < 24; index++) {
        CGFloat angle = index * M_PI / 12.0f;
        CGFloat currentRadius = index % 2 == 0 ? radius : innerRadius;

        [path addLineToPoint:CGPointMake(point.x + currentRadius * cos(angle),
                                         point.y + currentRadius * sin(angle))];
    }
}

+ (void)fourteenPointerStarAtPoint:(CGPoint)point
                            radius:(CGFloat)radius
                              path:(UIBezierPath *)path {
    CGFloat innerRadius = radius * 2.0f / 5.0f;
    CGFloat angleIncrement = M_PI * 2.0f / 28.0f;
    
    for (int index = 0; index < 28; index++) {
        CGFloat currentRadius = index % 2 == 0 ? radius : innerRadius;
        CGFloat angle = angleIncrement * index - M_PI / 2.0f;

        if (index == 0) {
            [path moveToPoint:(CGPointMake(point.x + radius * cos(angle), point.y + radius * sin(angle)))];
        } else {
            [path addLineToPoint:(CGPointMake(point.x + currentRadius * cos(angle), point.y + currentRadius * sin(angle)))];
        }
    }
}

+ (void)twentyFourPointerStarAtPoint:(CGPoint)point
                              radius:(CGFloat)radius
                         innerRadius:(CGFloat)innerRadius
                                path:(UIBezierPath *)path {
    [path moveToPoint:CGPointMake(point.x + radius * cos(0.0f),
                                  point.y + radius * sin(0.0f))];

    for (int index = 1; index < 48; index++) {
        CGFloat angle = index * M_PI / 24.0f;
        CGFloat currentRadius = index % 2 == 0 ? radius : innerRadius;

        [path addLineToPoint:CGPointMake(point.x + currentRadius * cos(angle),
                                         point.y + currentRadius * sin(angle))];
    }
}

@end

File: Sources/CountriesFlags/Extensions/UIBezierPath+Star.swift
//
//  UIBezierPath+Star.swift
//  CountriesFlags
//

import UIKit

extension UIBezierPath {

    enum StarType {
        case fourPointed
        case fivePointed
        case sixPointed
        case sevenPointed
        case eightPointed
        case twelvePointed(innerRadius: CGFloat)
        case fourteenPointed
        case twentyFourPointed(innerRadius: CGFloat)
    }

    static func star(at point: CGPoint, radius: CGFloat, type: StarType = .fivePointed) -> UIBezierPath {
        let path = UIBezierPath()

        path.move(to: CGPoint(x: point.x, y: -radius + point.y))

        switch type {
        case .fourPointed: fourPointerStar(at: point, with: radius, path: path)
        case .fivePointed: fivePointerStar(at: point, with: radius, path: path)
        case .sixPointed: sixPointerStar(at: point, with: radius, path: path)
        case .sevenPointed: sevenPointerStar(at: point, with: radius, path: path)
        case .eightPointed: eightPointerStar(at: point, with: radius, path: path)
        case .twelvePointed(let innerRadius): twelvePointedStar(at: point,
                                                                with: radius,
                                                                innerRadius: innerRadius,
                                                                path: path)
        case .fourteenPointed: fourteenPointedStar(at: point, with: radius, path: path)
        case .twentyFourPointed(let innerRadius): twentyFourPointed(at: point,
                                                                    with: radius,
                                                                    innerRadius: innerRadius,
                                                                    path: path)
        }

        path.close()
        return path
    }

    static private func fourPointerStar(at point: CGPoint,
                                        with radius: CGFloat,
                                        path: UIBezierPath) {
        let innerRadius = radius * 2.0 / 6.0
        for index in 1..<8 {
            let currentRadius = index % 2 == 0 ? radius : innerRadius
            let angle: CGFloat = -.pi / 2.0 + CGFloat(index) * .pi / 4.0

            let xPos: CGFloat = currentRadius * cos(angle)
            let yPos: CGFloat = currentRadius * sin(angle)
            path.addLine(to: CGPoint(x: point.x + xPos, y: point.y + yPos))
        }
    }

    static private func fivePointerStar(at point: CGPoint,
                                        with radius: CGFloat,
                                        path: UIBezierPath) {
        let theta: Float = 2.0 * Float.pi * 2.0 / 5.0
        for index in 1..<5 {
            let xPos: CGFloat = radius * CGFloat( sinf(Float(index) * theta) )
            let yPos: CGFloat = radius * CGFloat( cosf(Float(index) * theta) )
            path.addLine(to: CGPoint(x: point.x + xPos, y: point.y - yPos))
//                print("Five pointed star: x: \(xPos), y: \(yPos)")
        }
    }

    static private func sixPointerStar(at point: CGPoint,
                                       with radius: CGFloat,
                                       path: UIBezierPath) {
        var angle = -CGFloat.pi / 2.0
        for index in 0..<6 {
            if index == 0 {
                path.move(to: CGPoint(x: point.x + cos(angle) * radius,
                                      y: point.y + sin(angle) * radius))
            } else {
                path.addLine(to: CGPoint(x: point.x + cos(angle) * radius,
                                         y: point.y + sin(angle) * radius))
            }
            angle += CGFloat.pi / 6.0

            path.addLine(to: CGPoint(x: point.x + cos(angle) * (radius * 0.58),
                                     y: point.y + sin(angle) * (radius * 0.58)))
            angle += CGFloat.pi / 6.0
        }
    }

    static private func sevenPointerStar(at point: CGPoint,
                                         with radius: CGFloat,
                                         path: UIBezierPath) {
        let innerRadius = radius / 2.0 * 0.7
        let angleIncrement = .pi * 2.0 / 7.0

        let startAngle: CGFloat = -.pi / 2.0
        path.move(to: CGPoint(x: point.x + radius * cos(startAngle),
                              y: point.y + radius * sin(startAngle)))

        for index in 1..<14 {
            let angle = startAngle + angleIncrement * CGFloat(index)
            let currentRadius: CGFloat = index % 2 == 0 ? radius : innerRadius

            path.addLine(to: CGPoint(x: point.x + currentRadius * cos(angle),
                                     y: point.y + currentRadius * sin(angle)))
        }
    }

    static private func eightPointerStar(at point: CGPoint,
                                         with radius: CGFloat,
                                         path: UIBezierPath) {
        for index in 0..<16 {
            let angle: CGFloat = CGFloat(index) * .pi / 8.0
            let currentRadius = index % 2 == 0 ? radius : radius / 2.0

            if index == 0 {
                path.move(to: CGPoint(x: point.x + currentRadius * cos(angle),
                                      y: point.y + currentRadius * sin(angle)))
            } else {
                path.addLine(to: CGPoint(x: point.x + currentRadius * cos(angle),
                                         y: point.y + currentRadius * sin(angle)))
            }
        }
    }

    static private func twelvePointedStar(at point: CGPoint,
                                          with radius: CGFloat,
                                          innerRadius: CGFloat,
                                          path: UIBezierPath) {
        path.move(to: CGPoint(x: point.x + radius * cos(0.0),
                              y: point.y + radius * sin(0.0)))

        for index in 1..<24 {
            let angle = CGFloat(index) * .pi / 12.0
            let currentRadius: CGFloat = index % 2 == 0 ? radius : innerRadius

            let point = CGPoint(x: point.x + currentRadius * cos(angle),
                                y: point.y + currentRadius * sin(angle))
            path.addLine(to: point)
        }
    }

    static private func fourteenPointedStar(at point: CGPoint,
                                            with radius: CGFloat,
                                            path: UIBezierPath) {
        let innerRadius = radius * 2.0 / 5.0
        let angleIncrement = .pi * 2.0 / 28.0

        for index in 0..<28 {
            let currentRadius: CGFloat = index % 2 == 0 ? radius : innerRadius
            let angle: CGFloat = angleIncrement * CGFloat(index) - .pi / 2.0

            if index == 0 {
                path.move(to: CGPoint(x: point.x + radius * cos(angle),
                                      y: point.y + radius * sin(angle)))
            } else {
                path.addLine(to: CGPoint(x: point.x + currentRadius * cos(angle),
                                         y: point.y + currentRadius * sin(angle)))
            }
        }
    }
    
    static private func twentyFourPointed(at point: CGPoint,
                                          with radius: CGFloat,
                                          innerRadius: CGFloat,
                                          path: UIBezierPath) {
        path.move(to: CGPoint(x: point.x + radius * cos(0.0),
                              y: point.y + radius * sin(0.0)))

        for index in 1..<48 {
            let angle = CGFloat(index) * .pi / 24.0
            let currentRadius: CGFloat = index % 2 == 0 ? radius : innerRadius

            let point = CGPoint(x: point.x + currentRadius * cos(angle),
                                y: point.y + currentRadius * sin(angle))
            path.addLine(to: point)
        }
    }
}
